File: reverse_linked_list.js
/*在一个解构表达式中可以交换两个变量的值。
没有解构赋值的情况下，交换两个变量需要一个临时变量*/
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
/*cur,prev节点集体每次后移一位，每次改变的是next指针的指向*/
var reverseList = function(head) {
    let cur = head;
    let prev = null;
    while (cur) {
        [cur.next, prev, cur] = [prev, cur, cur.next]
    }
    return prev
};

File: reverse_linked_list.py
# reverse linked list
# 输入: 1->2->3->4->5->NULL
# 输出: 5->4->3->2->1->NULL
# 把每个结点的next指到它的前区节点上
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

'''class Solution(object):
    def reverseList(self, head):'''
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
def reverseList(self, head):
    cur, prev = head, None
    while cur:
        cur.next, prev, cur = prev, cur, cur.next
    return prev
